Refactor most of `kubectl drain` as a library

errordeveloper opened the pull request:
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Currently draining a node is only possible with `kubectl drain`. Other projects, e.g. [cluster autoscaler](https://github.com/kubernetes/autoscaler/blob/dadf2728eb17dc2da821a09c166b036b145163d3/cluster-autoscaler/utils/drain/drain.go) and [openshift](https://github.com/openshift/kubernetes-drain/blob/c2e51be1758efa30d71a4d30dc4e2db86b70a4df/drain.go) have reinvented some of this code. This PR moves what can be moved easily from `pkg/kubectl/cmd/drain` into `pkg/util/node/drain`.

This is intended as a simple code move, there are no changes to in behaviour.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

errordeveloper committed:
Refactor core functionality of `kubectl drain` as a library

- structured pod filter functions
- naming improvements
  - consistent use of daemonSets and DaemonSets
  - rename field to reflect its usage
- new cordon/uncordon helper
  - use Core API client direcly instead of generic CLI runtime

errordeveloper commented on pkg/util/node/drain/drain.go:
```diff
@@ -0,0 +1,279 @@
+		return []corev1.Pod{}, errors.New(fs.Message())
+	}
+	if len(ws) > 0 {
+		fmt.Fprintf(o.ErrOut, "WARNING: %s\n", ws.Message())
```
This is a little odd side-effect, I'll see how I can get rid of it easily.

errordeveloper commented on pkg/kubectl/cmd/drain/drain.go:
```diff
@@ -640,7 +430,7 @@ func (o *DrainOptions) deletePods(pods []corev1.Pod, getPodFn func(namespace, na
 }
 
-func (o *DrainOptions) waitForDelete(pods []corev1.Pod, interval, timeout time.Duration, usingEviction bool, getPodFn func(string, string) (*corev1.Pod, error)) ([]corev1.Pod, error) {
+func (o *DrainCmdOptions) waitForDelete(pods []corev1.Pod, interval, timeout time.Duration, usingEviction bool, getPodFn func(string, string) (*corev1.Pod, error)) ([]corev1.Pod, error) {
```
This function wants to write output a lot, I've not moved it cause I believe ideally libraries shouldn't write output. I've not figured out how alternative version of this would look like, and decide to not do that for the purpose of first PR.

neolit123 replied:
backends shouldn't be necessary on mute; they should allow piping to a custom stream.

i see this is already possible using `o.Out/ErrOut`, yet if this is always printing to the standard streams and if there is no easy way to customize it, it needs a minor refactor.
but given `o.Out/ErrOut` exists i'd consider moving this function too.

errordeveloper replied:
Yes, but the streams are `genericclioptions.IOStreams`, and those are not convenient to use outside of Kubernetes project itself, as an additional wrapping is needed to attach those to a logger or something else.

neolit123 replied:
i see, so that's a problem, yes.
consumers should have an easy way to manage the streams.

errordeveloper replied:
Yes, so I'm think of addressing this in a follow-up PR, what do you think?

neolit123 replied:
sounds good!

errordeveloper commented:
/test pull-kubernetes-integration

wking commented:
It's going to take me a bit to work through the diff of this vs. [openshift/kubernetes-drain][1], but just comparing the APIs, it looks like you've decided to not include (un)cordon functions?  Are you expecting to do that in follow-up work?  It seems like they'd be important for folks trying to use a drain library.

[1]: https://godoc.org/github.com/openshift/kubernetes-drain

errordeveloper commented:
> APIs, it looks like you've decided to not include (un)cordon functions? Are you expecting to do that in follow-up work? 

Yes, indeed I am intending to do that. I also need those myself, of course.

errordeveloper commented:
> It's going to take me a bit to work through the diff of this vs. openshift/kubernetes-drain

Have you tried to compare it to what's in master here in k/k now?

wking commented:
> > It's going to take me a bit to work through the diff of this vs. openshift/kubernetes-drain
>
> Have you tried to compare it to what's in master here in k/k now?

From openshift/kubernetes-drain@2b21795ef, you can see that I split off from the then-current d43e1b3.  Changes since then:

```console
$ git describe --always origin/master
v1.14.0-alpha.0-1656-gdc6f3d6
$ git log --oneline --decorate d43e1b3..dc6f3d6 -- pkg/kubectl/cmd/drain.go
33adf36 Move each kubectl command to a separate directory
97b2992 Update gofmt for go1.11
fecb5ed Merge pull request #66266 from wking/kubectl-drain-drop-backOff
686f29f (origin/pr/69438) Merge pull request #66301 from wking/kubectl-drain-drop-typer
f008365 Merge pull request #68806 from seans3/legacy-scheme-update
ab993e3 (origin/pr/66266) kubectl: Drop backOff from DrainOptions
7a2a987 Move legacyscheme (internal version) to kubectl scheme (external version)
452615c (origin/pr/68767) Fix drain for evicting terminal DS pods and pods with local storage
967280b (origin/pr/68069) Add --server-dry-run flag to `kubectl apply`
5b55e1f (origin/pr/67658) Create cli-runtime staging repository
0d81e83 (origin/pr/66301) kubectl: Drop typer from DrainOptions
```

#66266 and #66301 ported some of my downstream changes back upstream here.  I'll take a close look at #68767, #68806, and the others when I get a moment.

errordeveloper commented:
I am really not sure why this error appears in tests:
```
            error: unable to cordon node "node": no kind "Node" is registered for the internal version of group "" in scheme "pkg/kubectl/scheme/scheme.go:28"
```

I assume it breaks tests as output has changed, but I worry that the above error was there before this change.

neolit123 commented:
/priority important-longterm

neolit123 commented:
i think this can be fixed by just calling ./hack/update-bazel.sh

```
W0122 13:20:21.363] GoCompile: missing strict dependencies:
W0122 13:20:21.363] 	/bazel-scratch/.cache/bazel/_bazel_root/e9f728bbd90b3fba632eb31b20e1dacd/sandbox/linux-sandbox/2345/execroot/__main__/pkg/kubectl/cmd/drain/drain.go: import of "k8s.io/client-go/rest"
```

neolit123 commented on pkg/util/node/drain/filters.go:
```diff
@@ -0,0 +1,218 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
2019

neolit123 commented on pkg/util/node/drain/filters.go:
```diff
@@ -0,0 +1,218 @@
+	return makePodDeleteStatusWithWarning(false, daemonsetWarning)
+}
+
+func mirrorPodFilter(pod corev1.Pod) podDeleteStatus {
```
for consistency, i think it would make sense to still make this function as a member of `Options` even if it does not use the object fields.

errordeveloper replied:
👍

neolit123 commented on pkg/util/node/drain/filters.go:
```diff
@@ -0,0 +1,218 @@
+	// management resource - including DaemonSet - is not found).
+	// Such pods will be deleted if --force is used.
+	controllerRef := getPodController(pod)
+	if controllerRef == nil || controllerRef.Kind != "DaemonSet" {
```
non-blocking, nit as this is a copy-paste from the old code, but this should have been a comparison against a GVK -> Kind instead of a comparison to a string.

neolit123 commented on pkg/util/node/drain/drain.go:
```diff
@@ -0,0 +1,168 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
2019

neolit123 commented on pkg/util/node/drain/drain.go:
```diff
@@ -0,0 +1,168 @@
+// Options contains the parameters to control the behaviour of drainer
+type Options struct {
+	Client kubernetes.Interface
+	// RESTClient         *restclient.RESTClient
```
would it be safe to remove this field at this point?

errordeveloper replied:
Yes, indeed.

neolit123 reviewed (commented):
thanks for the updates @errordeveloper 
added some more comments.

errordeveloper commented:
/retest

On Tue, 12 Feb 2019, 2:27 pm Kubernetes Prow Robot <notifications@github.com
wrote:

> @errordeveloper <https://github.com/errordeveloper>: The following tests
> *failed*, say /retest to rerun them all:
> Test name Commit Details Rerun command
> pull-kubernetes-e2e-kops-aws 66fdbd5
> <https://github.com/kubernetes/kubernetes/commit/66fdbd5f23f98d23fcff04a2332a84e53352764b>
> link
> <https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/72827/pull-kubernetes-e2e-kops-aws/120254/> /test
> pull-kubernetes-e2e-kops-aws
> pull-kubernetes-bazel-test a542df8
> <https://github.com/kubernetes/kubernetes/commit/a542df8fb21e9b65ea5969b963f8ba9cc5913feb>
> link
> <https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/72827/pull-kubernetes-bazel-test/75581/> /test
> pull-kubernetes-bazel-test
> pull-kubernetes-node-e2e a542df8
> <https://github.com/kubernetes/kubernetes/commit/a542df8fb21e9b65ea5969b963f8ba9cc5913feb>
> link
> <https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/72827/pull-kubernetes-node-e2e/119496/> /test
> pull-kubernetes-node-e2e
>
> Full PR test history <https://gubernator.k8s.io/pr/72827>. Your PR
> dashboard <https://gubernator.k8s.io/pr/errordeveloper>. Please help us
> cut down on flakes by linking to
> <https://git.k8s.io/community/contributors/devel/flaky-tests.md#filing-issues-for-flaky-tests>
> an open issue
> <https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open> when
> you hit one in your PR.
>
> Instructions for interacting with me using PR comments are available here
> <https://git.k8s.io/community/contributors/guide/pull-requests.md>. If
> you have questions or suggestions related to my behavior, please file an
> issue against the kubernetes/test-infra
> <https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:>
> repository. I understand the commands that are listed here
> <https://go.k8s.io/bot-commands>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/72827#issuecomment-462779714>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPWS2mU409SnCGat5bivXZRE0TIO1fUks5vMs9ggaJpZM4Z7wev>
> .
>

neolit123 commented:
/skip
we were instructed to add skip to PRs that are blocking on `pull-kubernetes-e2e-kops-aws`
it was removed from presubmits, so old PRs might be stuct still trying to test it.

errordeveloper commented:
I've fix two bugs and the unit tests are passing locally now.

errordeveloper commented:
This is getting into a better shape now, @neolit123 it'd be great if you could take another look! There is one issue right now that comes up in manual tests - duplicate output lines.

errordeveloper commented:
Unit tests seem to panic when ran via bazel... hm!

errordeveloper commented:
/retest

errordeveloper commented:
I've fixed the race condition that bazel tests detected, which was to do with deletions lists containing non-unique items. I've also added an assertion in the tests that should prevent same from happening again.

errordeveloper commented:
If the tests pass in CI, I'm happy with this as it stands and would like to hear what reviewers/approvers think of this.

errordeveloper commented:
/retest

neolit123 reviewed (approved):
LGTM on my side @errordeveloper 
thank you!

errordeveloper commented:
cc @kubernetes/sig-cli-maintainers

errordeveloper commented:
/assign @pwittrock

soltysh reviewed (changes requested):
/hold
This PR is destroying the progress sig-cli has been going to split kubectl out of main repo. For that we CANNOT rely on any parts from the main repo. You can copy the functionality but you cannot make kubectl rely on any package from within k/k repo.

errordeveloper commented:
Hi @soltysh! I certainly had no intention of destroying any kind of progress. Let's find a better approach to this, I'm happy to connect here or on Slack. And if you insist, in a SIG meeting otherwise.

errordeveloper commented:
/assign @soltysh

As we discussed on Slack, I've moved the new package under `pkg/kubectl/drain`. If there is anything else that I should do, please let me know. The PR still only does a partial refactoring, and it seems plausible to finish it all up in a follow-up PR. With regards to tests, there aren't any tests yet for the new package, it's only covered by existing unit tests in `pkg/kubectl/cmd/drain` and whatever e2e tests there that cover `kubectl drain` functionality as such.

errordeveloper commented:
/retest

On Tue, 19 Feb 2019, 6:17 pm Kubernetes Prow Robot <notifications@github.com
wrote:

> @errordeveloper <https://github.com/errordeveloper>: The following tests
> *failed*, say /retest to rerun them all:
> Test name Commit Details Rerun command
> pull-kubernetes-e2e-kops-aws 66fdbd5
> <https://github.com/kubernetes/kubernetes/commit/66fdbd5f23f98d23fcff04a2332a84e53352764b>
> link
> <https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/72827/pull-kubernetes-e2e-kops-aws/120254/> /test
> pull-kubernetes-e2e-kops-aws
> pull-kubernetes-e2e-gce 37abbf0
> <https://github.com/kubernetes/kubernetes/commit/37abbf0f3fa811b5425f80d4effe464180df35ed>
> link
> <https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/72827/pull-kubernetes-e2e-gce/69228/> /test
> pull-kubernetes-e2e-gce
> pull-kubernetes-verify d7faf9d
> <https://github.com/kubernetes/kubernetes/commit/d7faf9d8310ca949a598a6354bd628bf8d3a0425>
> link
> <https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/72827/pull-kubernetes-verify/122915> /test
> pull-kubernetes-verify
>
> Full PR test history <https://gubernator.k8s.io/pr/72827>. Your PR
> dashboard <https://gubernator.k8s.io/pr/errordeveloper>. Please help us
> cut down on flakes by linking to
> <https://git.k8s.io/community/contributors/devel/flaky-tests.md#filing-issues-for-flaky-tests>
> an open issue
> <https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open> when
> you hit one in your PR.
>
> Instructions for interacting with me using PR comments are available here
> <https://git.k8s.io/community/contributors/guide/pull-requests.md>. If
> you have questions or suggestions related to my behavior, please file an
> issue against the kubernetes/test-infra
> <https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:>
> repository. I understand the commands that are listed here
> <https://go.k8s.io/bot-commands>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/72827#issuecomment-465248675>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPWS9Lm0JPvfUdQwNBLE4l5T8dlsRn3ks5vPD_IgaJpZM4Z7wev>
> .
>

soltysh commented:
/hold cancel

soltysh reviewed (approved):
/approve

soltysh commented:
Please squash your changes and I'll push that forward.

awh commented:
Nice work @errordeveloper - once I can use this in `kured` I no longer have to include `kubectl` in the image, which will reduce its size significantly.

soltysh reviewed (approved):
/lgtm

tedyu commented on pkg/kubectl/cmd/drain/drain.go:
```diff
@@ -344,218 +304,43 @@ func (o *DrainOptions) RunDrain() error {
-		if podOk {
-			pods = append(pods, pod)
+		if newErrs != nil {
+			fmt.Fprintf(o.ErrOut, "following errors also occurred:\n%s", utilerrors.NewAggregate(newErrs))
```
Should newErrs be aggregated into err ?

errordeveloper replied:
You mean from the library call? That can be discussed. There are still quite a few improvements to be made.

tedyu replied:
What do you think of the following ?
```
diff --git a/pkg/kubectl/cmd/drain/drain.go b/pkg/kubectl/cmd/drain/drain.go
index 8e98e84ee6..76beaef2cf 100644
--- a/pkg/kubectl/cmd/drain/drain.go
+++ b/pkg/kubectl/cmd/drain/drain.go
@@ -322,6 +322,8 @@ func (o *DrainCmdOptions) deleteOrEvictPodsSimple(nodeInfo *resource.Info) error
                }
                if newErrs != nil {
                        fmt.Fprintf(o.ErrOut, "following errors also occurred:\n%s", utilerrors.NewAggregate(newErrs))
+                       newErrs = append(newErrs, err)
+                       err = utilerrors.NewAggregate(newErrs)
                }
                return err
        }
```

errordeveloper replied:
It could be good, but I am not entirely sure, as the intention here was to avoid drastic changes of kubectl output, or general behaviour. This code originally returned the first error and printed new errors, so I kept it the same way. Not say it has to be that way, please open another PR if you strongly believe this must change. Small PR would be easy to discuss.
